CRM/Admin/Form/PaymentProcessor.php

```diff
@@ -400,4 +401,34 @@ public function updatePaymentProcessor(&$values, $domainID, $test) {
     civicrm_api3('PaymentProcessor', 'create', $params);
   }
 
+  /**
+   * Check if there is a record with the same name in the db.
+   *
+   * @param string $value
+   *   The value of the field we are checking.
+   * @param array $options
+   *   The daoName and fieldName (optional ).
+   *
+   * @return bool
+   *   true if object exists
+   */
+  public static function paymentProcessorNameExists($value, $options) {
+    $fieldName = 'name';
+    $daoName = CRM_Utils_Array::value(0, $options);
```

Comment:
I kept this rule local, although with the idea it could maybe be used by something, just to fix the issue at hand.  The nameExists rule already exists and is used quite a bit.  I'm not sure if limiting it by domain by default would effect other uses of it.  As domain id is not included in every entity.  Personally I like the idea of passing in the domain id for now and modifying the current nameExists rule to handle that.

As for use of the API by core.  I think someone needs to decide the rules about that.  I'd imagine you may want to look ahead at a sort headless civicrm.  I.e. only have UI code use the api not Core.  Granted this is UI code so maybe it works.  But absent guidelines rather then try to start that here I think it's best to stick with how all the other rules are written and follow that format.  Then when you come to a decision someone can do a PR which modifies all the rules to use the API.